pkg/util/backoff.go

```diff
@@ -0,0 +1,63 @@
+package util
+
+import (
+	"math/rand"
+	"time"
+)
+
+type BackoffConfig struct {
+	MinBackoff time.Duration // start backoff at this level
+	MaxBackoff time.Duration // increase exponentially to this level
+	MaxRetries int           // give up after this many; zero means infinite retries
+}
+
+type Backoff struct {
+	cfg        BackoffConfig
+	done       <-chan struct{}
+	numRetries int
+	cancelled  bool
+	duration   time.Duration
+}
+
+func NewBackoff(cfg BackoffConfig, done <-chan struct{}) *Backoff {
+	return &Backoff{
+		cfg:      cfg,
+		done:     done,
+		duration: cfg.MinBackoff,
+	}
+}
+
+func (b *Backoff) Reset() {
+	b.numRetries = 0
+	b.cancelled = false
+	b.duration = b.cfg.MinBackoff
+}
+
+func (b *Backoff) Finished() bool {
```

Comment:
Only minor comment: everywhere this is it used, the meaning if inverted (we're asking NotFinished).  Perhaps inverting this function would make a slightly nicer interface?  Call it `Remaining` or something?